Add token validation

JoJoPuppe opened the pull request:
I added token validation to the lifesaver middleware. Testing showed that **revoked** and **invalid** tokens return always a **404** status code. i am not sure if this is intended. i updated the scoring to punish also **404** errors.

JoJoPuppe committed:
Add token validation to lifesafer

stanislav-zaprudskiy commented on cc/keystone/middleware/lifesaver.py:
```diff
@@ -145,12 +145,70 @@ def get_user(self, request):
+                if 'auth' in body:
+                    if 'identity' in body['auth']:
+                        if 'token' in body['auth']['identity']:
+                            token_id = body['auth']['identity']['token'].get('id', None)
```
How about
> Flat is better than nested.

?

stanislav-zaprudskiy commented on cc/keystone/middleware/lifesaver.py:
```diff
@@ -145,12 +145,70 @@ def get_user(self, request):
+        token_id = None
+
+        try:
+            # grab toekn from an authentication request
```
```suggestion
            # grab token from an authentication request
```

bbobrov replied:
s/toekn/token/

bbobrov commented:
This should go to branch "master" first

bbobrov commented on cc/keystone/middleware/lifesaver.py:
```diff
@@ -145,12 +145,70 @@ def get_user(self, request):
+                if 'auth' in body:
+                    if 'identity' in body['auth']:
+                        if 'token' in body['auth']['identity']:
+                            token_id = body['auth']['identity']['token'].get('id', None)
```
Here the following case is covered. A user has a token. The user takes the token and wants obtain a new one, using the existing one to authenticate themselves. This might be needed to re-scope the token, or might be poorly automated by various scripts that perform the re-authentication if provided with a token. This is good, but this is not the only case that needs to be covered.

Another case is: a user, owning a correct token, wants to verify another token. They make a GET request to /v3/auth/tokens and pass their token in X-Auth-Token and the token they want to verify in X-Subject-Token. This is typically done by other openstack services, like nova or neutron. And often people are doing weird things and spam nova with requests with bad tokens. Nova has nothing else as to spam us.

We want to catch the bad tokens in X-Subject-token and limit them based on this logic.

JoJoPuppe replied:
done

stanislav-zaprudskiy replied:
It could be (and is likely) using `HEAD` for checking - https://docs.openstack.org/api-ref/identity/v3/#check-token. However, `GET` could still be relevant in the context of the problem too.

JoJoPuppe replied:
In the workgroup meeting, we decided to add routes and HTTP methods that appear in incidents for now. so we extend as we need it.

bbobrov commented on cc/keystone/middleware/lifesaver_utils.py:
```diff
@@ -48,7 +48,7 @@ def __init__(self, conf):
             group=group)
         CONF.register_opt(
-            cfg.DictOpt('status_cost', default=conf.get('status_cost', "default:1,401:10,403:5,404:0,429:0"),
+            cfg.DictOpt('status_cost', default=conf.get('status_cost', "default:1,401:10,403:5,404:10,429:0"),
```
Sorry, but this is a no-go. We cannot not generally limit users for their 404 requests. 404 requests are numerous, and normal, and are typically generated during a normal workflow. A single loop with `openstack` command will cause many 404s. There must be another way.

JoJoPuppe replied:
ok, then i need to see if i find another metric to check if the token is invalid.

JoJoPuppe replied:
so this was a [change](https://opendev.org/openstack/keystone/commit/53a6dc0e6a95f74905106d51f650bf4c2014bc08) 10 years ago. where most of the 401 Status codes where changed to 404 not found.

i get 401 for the X-Subject-Token flow, but revoked, invalid and token format errors i get 404. i added a separate token cost config.

bbobrov reviewed (changes requested):
The comments are inline.

stanislav-zaprudskiy commented on cc/keystone/middleware/lifesaver.py:
```diff
@@ -145,12 +145,70 @@ def get_user(self, request):
+                        if 'token' in body['auth']['identity']:
+                            token_id = body['auth']['identity']['token'].get('id', None)
+
+        except Exception as e:
```
`Exception` is very generic and would catch any error including, network hiccup, AttributeError or KeyboardInterrupt, for example. Can it be more specific here?

JoJoPuppe replied:
i added a Json decode exception.

JoJoPuppe committed:
Add X-Subject-Token validation

JoJoPuppe committed:
Add own token cost config

JoJoPuppe committed:
Change usage of reserved keyword type

JoJoPuppe committed:
Add specific Exception to token extraction

stanislav-zaprudskiy commented on cc/keystone/middleware/lifesaver_utils.py:
```diff
@@ -63,21 +66,24 @@ def __init__(self, conf):
+            user = user.decode("utf-8")
         return hashlib.md5(user.encode()).hexdigest()
 
-    def get_user_score(self, user):
```
It's not renamed everywhere, there are still places utilizing the old name.

stanislav-zaprudskiy replied:
Still the case

JoJoPuppe replied:
was missing the cli part. thanks for pointing this out.

stanislav-zaprudskiy commented on cc/keystone/middleware/lifesaver.py:
```diff
@@ -145,12 +149,79 @@ def get_user(self, request):
 
         return {'user': self.utils.normalize(user), 'domain': self.utils.normalize(domain)}
```
`get_user` handles password auth, ec2 auth, s3 auth and application credentials auth, returning a uniformly formatted dict. It may be reasonable to add token validation logic (currently added as `get_token` method) there, where the rest of the related logic is - and then calculate the costs based on the prefix of the returned `user` key in dict.

If this turns out a reasonable option, then `self.utils.normalize` might be not desirable here against tokens - as for tokens the case is important.

stanislav-zaprudskiy replied:
`normalize` should not be an issue if the token hash is utilized.

JoJoPuppe replied:
I extracted all authentication extraction methods into a separate file, and everything is treated equally now.

stanislav-zaprudskiy commented on cc/keystone/middleware/lifesaver.py:
```diff
@@ -145,12 +149,79 @@ def get_user(self, request):
+                token_id = token.get('id', None)
+            elif '/v3/auth/tokens' == request.path and 'GET' == request.method:
+                if "X-Subject-Token" in request.headers.keys():
+                    token_id = request.headers.get("X-Subject-Token")
```
A token could be
* valid
* invalid (e.g. never issued, or empty)
* expired
* revoked

And such tokens could appear in
* in an request authentication payload
  * `X-Auth-Token` header
  * or in body
* and/or in `X-Subject-Token` header

It's not clear if the code handles all reasonable combinations of those tokens - does it?

JoJoPuppe replied:
We tackle the use cases as needed. For now, we are adding POST and GET routes with the X-Subject-Token.

stanislav-zaprudskiy commented on cc/keystone/middleware/lifesaver.py:
```diff
@@ -145,12 +149,79 @@ def get_user(self, request):
+                identity = auth.get('identity', {})
+                token = identity.get('token', {})
+                token_id = token.get('id', None)
+            elif '/v3/auth/tokens' == request.path and 'GET' == request.method:
```
If I understand it right, potentially any Keystone API endpoint might require/benefit from 404 throttling, but we only enable throttling where we observe it's most needed - which is `POST /v3/auth/tokens` AFAIK.

The question is about this `GET/HEAD /v3/auth/tokens` then - do we just anticipate it's needed, or see it's needed? In the former case, why is it more likely to appear than e.g. any other API endpoint such as `GET /v3/users/{user_id}`?

If it's not immediately needed, we could reconsider to not add it, so to keep the code change smaller.

JoJoPuppe replied:
correct. we try now to keep changes small and iterate as needed.

stanislav-zaprudskiy commented on cc/keystone/middleware/lifesaver.py:
```diff
@@ -145,12 +149,79 @@ def get_user(self, request):
+            if item_score.refill_amount != self.utils.refill_amount:
+                item_score.refill_amount = self.utils.refill_amount
+
+            self.utils.set_score(item, item_score)
```
Not to store a sensitive token in memcache, it might be reasonable to store and compare checksum of it instead.

JoJoPuppe replied:
token are not stored anymore. every auth extraction method extracts the user_id and/or domain

stanislav-zaprudskiy commented on cc/keystone/middleware/lifesaver.py:
```diff
@@ -145,12 +149,79 @@ def get_user(self, request):
+                item_score.refill_amount = self.utils.refill_amount
+
+            self.utils.set_score(item, item_score)
+            self.logger.info("%s has a remaining credit of %d - request %s %s returned %d" % (
```
This would effectively store a token in logs. Are we sure that valid tokens - https://github.com/sapcc/keystone-extensions/pull/10/files#r2556797581 - will never be stored?

JoJoPuppe replied:
thank you. tokens are not stored any more. see previous comment.

stanislav-zaprudskiy commented on cc/keystone/middleware/lifesaver.py:
```diff
@@ -145,12 +149,79 @@ def get_user(self, request):
+                item_score.refill_amount = self.utils.refill_amount
+
+            self.utils.set_score(item, item_score)
+            self.logger.info("%s has a remaining credit of %d - request %s %s returned %d" % (
```
When seeing logs like that with token/token hash, what useful information could be extracted from such a message?

Perhaps for tokens we should consider logging some other context that could help with troubleshooting.

stanislav-zaprudskiy reviewed (commented):
Let's please discuss offline if and how we want to proceed with this PR before anyone spends any more time on it.

JoJoPuppe committed:
Put auth method extractions into separate file to reduce mocking for testing logic.

- remove token separation
- every extraction of auth credentials tries to get user_id and domain.
- with this approach we have no issues with logging token ids.

JoJoPuppe committed:
Add unit tests for extracting credentials and calculating costs in lifesaver_logic

JoJoPuppe committed:
Remove redundant tests for returning None in credential extraction

JoJoPuppe committed:
Add End-to-End tests for Lifesaver middleware and environment setup

JoJoPuppe commented:
The latest changes introduce unit tests for the extraction logic. These are now in a separate file to work around mocking all the Keystone relevant classes and methods.  
I also added end-2-end tests for each extraction method and for the refill process. This involves executing against a real Keystone instance, and I added detailed documentation on how to set this up. The end-2-end tests only represent the happy path of triggering the rate limit for each extraction method.

stanislav-zaprudskiy commented on cc/keystone/middleware/lifesaver.py:
```diff
@@ -58,99 +67,80 @@ def get_user(self, request):
 
         user = None
         domain = None
+        token_id = None
```
> Local variable 'token_id' value is not used

stanislav-zaprudskiy commented on tests/E2E_TESTING_README.md:
```diff
@@ -0,0 +1,349 @@
+# Running End2End LifeSaver Tests
+
+In order to be able to run those tests you need to have a local keystone instance running on your machine. the following steps will describe how to set everything up.
```
I'd think that Keystone setup procedure belongs elsewhere - for example, sapcc/keystone repo.

JoJoPuppe replied:
i have created this PR
https://github.com/sapcc/keystone/pull/46

JoJoPuppe committed:
Remove unused token_id variable from LifesaverMiddleware

JoJoPuppe committed:
Rename method get_user_score to get_score that was still used in cli

JoJoPuppe committed:
Update E2E testing instructions.

stanislav-zaprudskiy commented on tests/E2E_TESTING_README.md:
nitpick. Just `README.md` for this file would be completely fine.

stanislav-zaprudskiy reviewed (dismissed):
Want to see a demo :) but otherwise preliminary approved 👍

JoJoPuppe committed:
Refactor Lifesaver E2E tests.

- use domain IDs and update test admin password
- flush memcache before each test to make tests independent

JoJoPuppe committed:
refactor token extraction to hash the actual token for reference in memcache

reason being is that not only revoked tokens are invalid, also expired and
malformed tokens should be rate limited. and from invalid tokens we cannot
extract user or domain. so the compromise is we do not know how is sending
the invalid requests but we catch all the cases this way.
- since we do not extract for all scenarios the user, but also the token,
the variable user is changed to subject

JoJoPuppe committed:
add unit tests for hash_token_id function to ensure consistent and correct hashing behavior

JoJoPuppe committed:
refactor e2e test

refactor e2e test to take into account that invalid credentials will trigger
400+ errors we do not care about the validity. we only see the status code.
so this should be enough to test correct behavior.

JoJoPuppe commented:
in recent changes i introduced token hashing to track invalid tokens in memcache and their rate limiting.
also reduced the e2e to make the assumtion we are only seeing the status code from keystone request and making our decision based on this if we rate limit or not. so expired, invalid or revoked tokens should all give 400+ errors.

bbobrov commented on cc/keystone/middleware/lifesaver_utils.py:
```diff
@@ -63,21 +67,33 @@ def __init__(self, conf):
+
+    def hash_token_id(self, token_id: str) -> str:
+        secret_key = CONF.security_compliance.invalid_password_hash_secret_key
+        configured_hash_function = CONF.security_compliance.invalid_password_hash_function
```
We cannot expect that this is enabled everywhere. Keystone should still somehow work with these options disabled or not set.

bbobrov replied:
Probably it just should not block by token if the token hashing function is not available.

bbobrov replied:
Actually, i don't exactly understand how this all will work if not configured. How?

JoJoPuppe replied:
i decided to push a log warning if the config is not set and use the hostname as fallback, so this will never raise an Value Error and fail.

tz3 replied:
> Actually, i don't exactly understand how this all will work if not configured. How?

already handled `hash_token_id` in checks `if not secret_key` logs a warning, and falls back to `socket.getfqdn()` as the HMAC key.

https://github.com/sapcc/keystone-extensions/pull/10/changes#diff-ae3339250161585c3719fd0431ee51a9b5bd831478137d14bedf263390da152aR210-R227

bbobrov commented on cc/keystone/middleware/lifesaver.py:
```diff
@@ -45,112 +52,85 @@ def __init__(self, app, conf):
+        # determine cost based on response status
+        if status >= 400:
+            # check prefix to decide which cost table to use
+            if subject.startswith("FTOKENCREDS-"):
```
It is uppercase here and lowercase in another part of the code. It would be better to move these subjects into separate "objects", like this:
```
FTOKENCREDS = 'ftokencreds-'
...
if subject.startswith(FTOKENCREDS):
```

bbobrov commented on cc/keystone/middleware/lifesaver_logic.py:
```diff
@@ -0,0 +1,276 @@
+    ).hexdigest()
+
+
+def calculate_cost(status_code, user_identifier, status_cost_config, token_cost_config):
```
This seems to be never called

JoJoPuppe replied:
this is called here:
https://github.com/sapcc/keystone-extensions/blob/7c48f533ea952ba38d94424fec16fc351c38e79e/cc/keystone/middleware/lifesaver.py#L103-L109

bbobrov commented on cc/keystone/middleware/lifesaver_logic.py:
```diff
@@ -0,0 +1,276 @@
+    return int(cost_table.get('default', 1))
+
+
+def should_update_score_metadata(score, current_credit, current_refill_time, current_refill_amount):
```
this seems to be never called

JoJoPuppe replied:
it is called here:
https://github.com/sapcc/keystone-extensions/blob/7c48f533ea952ba38d94424fec16fc351c38e79e/cc/keystone/middleware/lifesaver.py#L116-L119

bbobrov commented on cc/keystone/middleware/lifesaver_logic.py:
```diff
@@ -0,0 +1,276 @@
+            return None, None
+
+        if request.path not in ['/v3/s3tokens', '/v3/ec2tokens']:
+            return None, None
```
this will never happen, because the path above is checked

bbobrov commented on cc/keystone/middleware/lifesaver.py:
```diff
@@ -160,67 +140,42 @@ def verify_request(self, request, response=None):
-                if user_score.get() == 0:
-                    self.logger.info("Blocking request %s %s, since user %s %s has no credit left" % (
-                    request.method, request.path, user, domain))
+                if subject_score.get() == 0:
```
i wonder - can it be negative?

tz3 replied:
score can't go negative, these two functiosn are the one responsible for the socre modification here are they:-

1. `reduce()` [line 46-47]( https://github.com/sapcc/keystone-extensions/blob/main/cc/keystone/middleware/score.py#L46-L48): ~if consumption > self.value`, it sets `self.value = 0` — never goes below zero.
2. `get(self)` [line 33-36 ](https://github.com/sapcc/keystone-extensions/blob/main/cc/keystone/middleware/score.py#L33-L36) returns min(credit, value + refills) — also never negative since value is clamped at 0

bbobrov committed:
Fix review issues: remove dead code, add FTOKENCREDS_PREFIX constant, wire up extracted functions

- Remove unused imports (JSONDecodeError, Request) from lifesaver.py
- Add FTOKENCREDS_PREFIX constant in lifesaver_logic.py and use it
  consistently across lifesaver.py, lifesaver_logic.py, and tests
- Refactor get_costs to call calculate_cost() and
  should_update_score_metadata() instead of duplicating logic inline
- Add explicit ValueError handling for hash_token_id when
  invalid_password_hash_secret_key is not configured
- Remove redundant path check in extract_s3_ec2_credentials

JoJoPuppe committed:
handle missing invalid_password_hash_secret_key gracefully

When invalid_password_hash_secret_key is not configured, token hashing
now falls back to the hostname as the HMAC secret and logs a warning,
instead of raising a ValueError and disabling token rate-limiting.

Also fixes the score blocking condition from == 0 to <= 0.

JoJoPuppe commented:
i rebased to the current base and also changed from a required config secret-key to a fall back using hostname.

bbobrov reviewed (commented):
<disregard this>

bbobrov reviewed (changes requested):
please see the next comment

bbobrov reviewed (commented):
Revised findings:

### High Severity

**1. Noisy info-level logging on every request** (`lifesaver.py:168,171`)

Two new `self.logger.info(...)` calls fire on the hot path — one in `process_request`, one in `process_response` — for every request with an identifiable subject. That's 2 extra info lines per request where the old code produced 0 on the happy path. In production this will flood logs and could impact performance. These are clearly debug statements that were left in. Should be `debug` level or removed.

### Medium Severity

**2. Dead guard in `extract_from_authentication_request`** (`lifesaver_logic.py:188`)

`if not user:` is always `True` since `user` was just set to `None` on line 182. This suggests the code was copied from the old monolithic method where `user` could already have a value from earlier extraction. In the new standalone function it's dead logic that gives a false impression that `user` might be pre-populated, making the function harder to reason about correctly.

JoJoPuppe committed:
feat(lifesaver): add type hints and refactor credential extraction logic

- Add type hint for subject_score variable
- Remove unnecessary debug logging statements
- Refactor extraction functions with early returns for path/method checks
- Reduce nesting depth in extract_password_auth_credentials and extract_app_credential

tz3 committed:
fix: preserve header user in extract_from_authentication_request and add token validation tests

- Fix bug where a user extracted from HTTP headers was silently
  overwritten by token_info when domain header was missing
- Correct unit test that was asserting the buggy overwrite behavior

tz3 reviewed (approved):
Thanks @JoJoPuppe for the pr, from my side looks good in the current state.

@bbobrov let me know if you have any concerns. I Rreviewed it, i pushed some changes.

I pushed [`commit 687c51b`](https://github.com/sapcc/keystone-extensions/commit/687c51b), because at [`1128093`](https://github.com/sapcc/keystone-extensions/commit/1128093) the user-overwrite bug existed but was silent — `test_falls_back_to_token_info_when_headers_incomplete` was asserting the buggy behavior so it passed. [`687c51b`](https://github.com/sapcc/keystone-extensions/commit/687c51b) fixes the code and corrects the test assertion to match the correct behavior

### What I tested

Tested locally against, I can confirm the following are covered:

Tested locally against a full Docker stack (keystone + memcached + mysql + openldap). I can confirm the following are covered:

- 31 unit tests in [`tests/test_lifesaver_logic.py`](https://github.com/sapcc/keystone-extensions/blob/add-token-validation/tests/test_lifesaver_logic.py) — all passing
- Full E2E suite in [`tests/test_lifesaver_e2e.py`](https://github.com/sapcc/keystone-extensions/blob/add-token-validation/tests/test_lifesaver_e2e.py) — all passing
- Manually verified: hash consistency, bucket isolation, zero cost for valid tokens, cost table separation (token 404 costs 10 vs status 404 costs 0), subject isolation between token and password buckets, unauthenticated validation (401 costs 10), and HEAD method bypass



---


**@bbobrov :**

> 1. Noisy info-level logging on every request

Removed in [`1128093`](https://github.com/sapcc/keystone-extensions/commit/1128093). Remaining `info` lines only fire on credit deduction, blocklist, and rate-limit block.

> 2. Dead guard `if not user:` always True

The guard at [`lifesaver_logic.py#L198`](https://github.com/sapcc/keystone-extensions/blob/add-token-validation/cc/keystone/middleware/lifesaver_logic.py#L198) is now intentional — the outer condition (`if not user or not domain:`) can enter with `user` already set when only `domain` is missing. Without it, the header user gets overwritten by token_info.

bbobrov commented:
This change has become un-reviewable.
It incorporates now 3 changes:
1. Tests
2. Refactoring
3. Token validation

This needs to be done differently. We need to split this into 3 separate PRs.

JoJoPuppe commented:
> This change has become un-reviewable. It incorporates now 3 changes:
> 
> 1. Tests
> 2. Refactoring
> 3. Token validation
> 
> This needs to be done differently. We need to split this into 3 separate PRs.

Thanks for the reply. I will prepare individual PRs.
